### PR TITLE
Ca 50 created scrape and scrape-meta for laserfichie

### PR DESCRIPTION
**Pull request by naumansharifwork**

### CODE REVIEW

**Comment by naumansharifwork** [comment]
Hey @zstumgoren please check this one as well it was a bit different because this website was dynamic its pages were loading after requesting api calls and when we download its .html files they have empty shells, so instead of working and downloading its html files to fetch the data from i have downloaded its data in jsons and fetched and created the meta file and using that meta file we are downloading the data further.

below is the json and sample folder structure

image
image

[ca_laserfiche.json](https://github.com/user-attachments/files/16429270/ca_laserfiche.json)

**Comment by tarakc02** [comment]
Hi @naumansharifwork I haven't taken a look at the rest of the code yet, but note the name for this agency is `fullerton`

**Comment by tarakc02** [comment]
@naumansharifwork,

- name of the script (and resulting agency slug) should be `fullerton_pd.py` (`fullerton_pd`) instead of `fullerton` (sorry about that)
- `parent_page` should be a relative path rather than absolute
- since `asset_url` is a required field: @zstumgoren in these cases with dynamic content, what do we do if there is no static `asset_url`?
- the calls to `cache.write_json` in line 78 and line 89 have the side effect of creating an empty `ca_fullerton` directory at the project root. as a workaround, this can be fixed within this script by passing `base_output_json` and `output_json` as the argument rather than `filename`, but @zstumgoren i believe the root cause here is [this line from cache.py](https://github.com/naumansharifwork/clean-scraper/blob/f8d0fc98917070c7cbb620bdc50f337136e95f4c/clean/cache.py#L197) referring to `out` rather than `fullpath` -- let me know if i should open a separate issue for that.
- when i try to run this locally, it downloads the first couple of files, and then lots of failures (finished: false), not sure if that's due to not waiting long enough between tries

**Comment by naumansharifwork** [comment]
@tarakc02 for the finished false also check the completed percentage, for the bigger files it shows finished as false for a long time until export is complete. Basically there are 3 requests that download the file,
1. Start the export ( it sends requests to server to download the file and server start compressing the file)
2. Checking export status ( this one is called repeatedly its response is the one that you are seeing. Basically it checks from the server if the file is ready to download or not when the complete percentage becomes 100 the finished becomes true and we move to next request.)
3. We get a token from the previous ones and we use this token to download the file.

**Comment by naumansharifwork** [comment]
@tarakc02 
updated export json
[ca_fullerton_pd.json](https://github.com/user-attachments/files/16439547/ca_fullerton_pd.json)

**Comment by tarakc02** [comment]
Thank you! This looks good to me for now. Awaiting @zstumgoren answer to the question about the `asset_url`, and also just to make sure the additional methods in `utils.py` are acceptable (they look good to me).

**Comment by zstumgoren** [comment]
> Thank you! This looks good to me for now. Awaiting @zstumgoren answer to the question about the `asset_url`, and also just to make sure the additional methods in `utils.py` are acceptable (they look good to me).

@naumansharifwork @tarakc02 I'm gonna hand you off to @newsroomdev who is developing some docs and a standardized schema. He's in the best position at this point to review and provide feedback until the docs/schema are published.

**Comment by newsroomdev** [comment]
Hi @naumansharifwork. I'm happy to review this pull request, but we'll need to update the code with the latest from the `dev` branch. Let me know if I can help with that. The added `utils` look great!

As for metadata JSON, I'd like to standardize these across scrapers (potentially refactoring existing scrapers) and would like to hear your thoughts on [an open GitHub Discussion](https://github.com/biglocalnews/clean-scraper/discussions/58). I don't think `fullerton_pd.json` matches just yet, but there may be additional distinct considerations we need to adjust for.

Lastly, @tarakc02 are you referring to a specific instance in this pull request where `asset_url` was a null-ish value? Basically, if the page is dynamic, the scraper would need to handle the site dynamically. This might be easier to discuss further via your [discussion comments](https://github.com/biglocalnews/clean-scraper/discussions/58#discussioncomment-10227382)

**Comment by naumansharifwork** [comment]
@newsroomdev i have updated the code from dev, for the case_num there is a case from the website which does not have the case_num the very first record on [https://portal.laserfiche.com/Portal/Browse.aspx?id=726681&repo=r-3261686e](url) so, should we keep it as empty?

**Comment by newsroomdev** [comment]
Interesting catch @naumansharifwork. Your example is a good reason to use `case_id` instead of `case_num`. The URL looks a little similar to San Diego PD's `case_id` structure where there's an address and an identification number. In this instance, we can safely default `case_id == title` but typically we'll want to extract a case_id from either the `parent_page`'s title or its contents.

```json
    {
        "title": "1 Status of Productions",
        "parent_page": "ca_fullerton_pd/1 Status of Productions.json",
        "asset_url": "https://portal.laserfiche.com/Portal/Browse.aspx?id=726681&repo=r-3261686e",
        "name": "Status of SB 1421 Productions"
        "case_id": "1 Status of Productions"
    },
```

File: clean/utils.py [hunk 1]
```diff
@@ -279,3 +279,61 @@ def get_repeated_asset_url(self, objects: List[MetadataDict]):
...
+        url (str): the url to be requested
+        user_agent (str): the user-agent header passed with the request (default: biglocalnews.org)
+        session: a session object to use when making the request. optional
+    """
+    logger.debug(f"Requesting {url}")
+
+    # Set the headers
+    if "headers" not in kwargs:
+        kwargs["headers"] = {}
+    kwargs["headers"]["User-Agent"] = user_agent
```
Comment (stucka) [inline clean/utils.py:300]: I think line 300 here might overwrite a User-Agent sent as part of the kwargs package, and that might be bad and break some scrapers that may need to imitate something else. An indent on line 300 would fix that, I think, but I don't know if it's desirable in all use cases.

**Comment by naumansharifwork** [comment]
@newsroomdev i have updated the script to look into child folders as well and get the asset_urls. 
attaching the output meta json as well.
please have a look Thanks.

[ca_fullerton_pd.json](https://github.com/user-attachments/files/16616847/ca_fullerton_pd.json)

**Pull request closed without merging**